liquibase-core/src/main/java/liquibase/Liquibase.java

```diff
@@ -285,6 +288,38 @@ public void update(Contexts contexts, LabelExpression labelExpression, boolean c
         });
     }
 
+    /**
+     * Performs check of the historyService to determine if there is no unrun changesets without obtaining an exclusive write lock.
+     * This allows multiple peer services to boot in parallel in the common case where there are no changelogs to run.
+     * <p>
+     * If we see that there is nothing in the changelog to run and this returns <b>true</b>, then regardless of the lock status we already know we are "done" and can finish up without waiting for the lock.
+     * <p>
+     * But, if there are changelogs that might have to be ran and this returns <b>false</b>, you MUST get a lock and do a real check to know what changesets actually need to run.
+     * <p>
+     * NOTE: to reduce the number of queries to the databasehistory table, this method will cache the "fast check" results within this instance under the assumption that the total changesets will not change within this instance.
+     */
+    protected boolean isUpToDateFastCheck(Contexts contexts, LabelExpression labelExpression) throws LiquibaseException {
+        String cacheKey = contexts +"/"+ labelExpression;
+        if (!this.upToDateFastCheck.containsKey(cacheKey)) {
```

Comment:
Not a big deal, but since Java 8, there's a "[computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-)" method on Map that lets you go a "get-or-put" op atomically without needing to do the map lookup twice:

```
return this.upToDateFastCheck.computeIfAbsent(cacheKey, k -> {
      return ... // compute uncached value here
})```

Reply:
Good point. I always forget computeIfAbsent saves the value that gets computed...

I'm good just leaving it as it is for now and we can do a general fixup of that pattern across the codebase at some point.